### ISSUE TITLE
Enable non-causal mask (to enable MLM) for VisionEncoderDecoder models

### ISSUE THREAD

**Issue by metemadi** [huggingface/transformers#24343]
### Feature request

Hello! The current (amazing!) VisionEncoderDecoder library supports text generation via a standard causal LM. Some recent work (linked [here](https://arxiv.org/abs/2306.07915)) has shown promise in having the text decoder be a MLM instead of a causal LM. I believe this is doable with the current VisionEncoderDecoder library by passing in [MASK] tokens for the decoder_input_ids and passing in the labels as usual, but this would still result in a causal mask. The code comment is as follows which makes me think this:
```
decoder_attention_mask (`torch.BoolTensor` of shape `(batch_size, target_sequence_length)`, *optional*):
            Default behavior: generate a tensor that ignores pad tokens in `decoder_input_ids`. Causal mask will also
            be used by default.
```
Is there a way to turn off causal masking to predict multiple text tokens at once using a VisionEncoderDecoder model?

### Motivation

Masked language modeling on top of a Vision encoder appears to be a promising new approach for image captioning and pre-training of vision models according to [this recent work](https://arxiv.org/abs/2306.07915).

### Your contribution

Thank you!

**Comment by amyeroberts** [comment]
Hi @metemadi, thanks for opening this issue! 

This sounds like an interesting project! I believe there's a few places that would need to be adapted in order to enable this properly, such as not forcing `add_cross_attention` to the decoder config and not shifting tokens (cc @ydshieh). The VisionEncoderDecoder model is not intended to be compatible with all encoder-decoder pairs or use cases. This isn't something we'll add to the library at the moment, but feel free to share a fork branch with an implementation here if you'd like!

**Comment by metemadi** [comment]
Thank you for the insanely fast reply - HuggingFace is amazing as always! This all makes sense. Thanks again.

**Comment by ydshieh** [comment]
Sorry, I forgot to reply:

There is however `class VisionTextDualEncoderModel`. One checkpoint on the Hub is [clip-italian](https://huggingface.co/clip-italian/clip-italian). If you look the config file, it uses `BertForMaskedLM` and `clip_vision_model`.

It might be helpful, but some slight modification might be necessary if the goal is to do what have been done in the paper you mentioned.
